### PR TITLE
resolve "user" and "workspace" debug configs

### PR DESCRIPTION
**Pull request by elahehrashedi**
bug fix; https://github.com/microsoft/vscode-cpptools/issues/9319

New Telemetry style:

**three telemetry event names:** 
- `debugPanel`   // F5 or "Run and Debug" debug panel
- `playButton`    // "Run and Debug" play button
- `addConfigGear`     // the gear icon

**debugType:**  `debug`, `run`

**configSource:** `singleFile`, `workspaceFolder`, `workspace`, `global`, `unknown`

**configMode:** `launchConfig`, `noLaunchConfig`

**cancelled**: `true`/`false`
**success**: `true`/`false`

### CODE REVIEW

File: Extension/src/Debugger/configurationProvider.ts [hunk 1]
```diff
@@ -37,36 +37,21 @@ export enum TaskConfigStatus {
...
-        if (item.configuration.taskDetail) {
-            // Add the compiler path of the preLaunchTask to the description of the debug configuration.
-            item.detail = (item.detail ?? "") + " (" + item.configuration.taskDetail + ")";
-        }
-
-    });
-    return items;
-};
+export enum DebugConfigSource {
+    singleFile = "singleFile",              // a debug config defiend for a single mode file
```
Comment (jogo-) [inline Extension/src/Debugger/configurationProvider.ts]: Typo (x4): defiend -> defined

Comment (elahehrashedi) [reply Extension/src/Debugger/configurationProvider.ts]: thank you @jogo-

File: Extension/src/Debugger/configurationProvider.ts [hunk 2]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+        /** If the config is coming from "Run and Debug" debugPanel, there are three cases where the folder is undefined:
+         * 1. when debugging is done on a single file where there is no folder open,
+         * 2. when the debug configuration is defined at the User level (global).
+         * 3. when the debug configuration is defined at the workspace level.
+         * If the config is coming from "Run and Debug" playButton, there is one case where the folder is undefined:
+         * 1. when debugging is done on a single file where there is no folder open.
+         */
+
+        /** Do not resolve PreLaunchTask for these two cases, and let the VsCode resolve it:
+         * 1: The configs that comes from playButton and are already defined, these configs should already have their PreLaunchTask defined.
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Missing word after "and" in the sentence "The configs that comes from playButton and are already defined, these configs should already have their PreLaunchTask defined."

File: Extension/src/Debugger/configurationProvider.ts [hunk 3]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+         * 2: The configs that comes from debugPanel where the folder is undefined but task can not be found.
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Nit: "can not" --> "cannot"

File: Extension/src/Debugger/configurationProvider.ts [hunk 4]
```diff
@@ -591,40 +577,115 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+            return matchingConfig.source;
+        }
+        return undefined;
+    }
+
+    public getLaunchConfigs(folder?: vscode.WorkspaceFolder, type?: DebuggerType | string): CppDebugConfiguration[] | undefined {
+        // Get existing debug configurations from launch.json or user/workspace "launch" settings.
+        const WorkspaceConfigs: vscode.WorkspaceConfiguration = vscode.workspace.getConfiguration('launch', folder);
+        const configs: any = WorkspaceConfigs.inspect<vscode.DebugConfiguration>('configurations');
+        // let configs: any = vscode.workspace.getConfiguration('launch', folder).get('configurations');
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: remove commented code on line 649?

File: Extension/src/Debugger/configurationProvider.ts [hunk 5]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
-            /* There are two cases where folder is undefined:
-                * when debugging is done on a single file where there is no folder open,
-                * or when the debug configuration is defined at the User level.
-                */
-            await this.resolvePreLaunchTask(undefined, config);
-            config.preLaunchTask = undefined;
-        } else {
-            await this.resolvePreLaunchTask(folder, config);
+
+        /** If the config is coming from "Run and Debug" debugPanel, there are three cases where the folder is undefined:
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: "from"->"from the"

File: Extension/src/Debugger/configurationProvider.ts [hunk 6]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+         * 1. when debugging is done on a single file where there is no folder open,
+         * 2. when the debug configuration is defined at the User level (global).
+         * 3. when the debug configuration is defined at the workspace level.
+         * If the config is coming from "Run and Debug" playButton, there is one case where the folder is undefined:
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: "from"->"from the"

File: Extension/src/Debugger/configurationProvider.ts [hunk 7]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+         * 1. when debugging is done on a single file where there is no folder open.
+         */
+
+        /** Do not resolve PreLaunchTask for these two cases, and let the VsCode resolve it:
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: "the VsCode"->"VS Code" or "vscode" -- the "the" would be correct if "VsCode" was an adjective describing another noun, such as "the VS Code implementation".

File: Extension/src/Debugger/configurationProvider.ts [hunk 8]
```diff
@@ -591,40 +577,115 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
         return false;
     }
 
+    private getDebugConfigSource(config: CppDebugConfiguration, folder?: vscode.WorkspaceFolder): DebugConfigSource | undefined {
+        if (config.source) {
+            return config.source;
+        }
+        const configs: CppDebugConfiguration[] | undefined = this.getLaunchConfigs(folder, config.type);
+        const matchingConfig: CppDebugConfiguration | undefined = configs?.find((item: any) => item.name && item.name === config.name);
+        if (matchingConfig && matchingConfig.source) {
```
Comment (michelleangela) [inline Extension/src/Debugger/configurationProvider.ts]: Simplify `if (matchingConfig && matchingConfig.source)` to `if (matchingConfig?.source)` ?

File: Extension/src/Debugger/configurationProvider.ts [hunk 9]
```diff
@@ -175,17 +130,48 @@ export class DebugConfigurationProvider implements vscode.DebugConfigurationProv
...
+            if (!debugPanelConfig && existingConfig) {
+                resolveByVsCode = true;
+            } else if (debugPanelConfig && !folder) {
+                isExistingTask = await cppBuildTaskProvider.isExistingTask(config.preLaunchTask);
+                if (!isExistingTask) {
+                    resolveByVsCode = true;
+                }
+            }
+
+            // Send Telemetry before writing into files.
```
Comment (sean-mcmanus) [inline Extension/src/Debugger/configurationProvider.ts]: "Telemetry"->"telemetry"

**Review by michelleangela — CHANGES_REQUESTED** [review]

File: Extension/src/LanguageServer/cppBuildTaskProvider.ts [hunk 10]
```diff
@@ -230,6 +230,19 @@ export class CppBuildTaskProvider implements TaskProvider {
         return this.writeBuildTask(taskLabel, true);
     }
 
+    public async isExistingTask(taskLabel: string, workspaceFolder?: WorkspaceFolder): Promise<any> {
+        const rawTasksJson: any = await this.getRawTasksJson(workspaceFolder);
+        if (!rawTasksJson.tasks) {
+            rawTasksJson.tasks = new Array();
+        }
+        // Check if the task exists in the user's task.json.
+        const selectedTask: any = rawTasksJson.tasks.find((task: any) => task.label && task.label === taskLabel);
```
Comment (sean-mcmanus) [inline Extension/src/LanguageServer/cppBuildTaskProvider.ts]: Can this be simplified to 1 return line, e.g. `return rawTasksJson...`?

**Review by sean-mcmanus — CHANGES_REQUESTED** [review]

**Review by michelleangela — APPROVED** [review]

**Review by sean-mcmanus — APPROVED** [review]

**Pull request merged**